core/services/workflows/engine.go

```diff
@@ -807,6 +858,95 @@ func (e *Engine) deregisterTrigger(ctx context.Context, t *triggerCapability, tr
 	return nil
 }
 
+func (e *Engine) isWorkflowFullyProcessed(state store.WorkflowExecution) (bool, string, error) {
+	statuses := map[string]string{}
+	// we need to first propagate the status of the errored status if it exists...
+	err := e.workflow.walkDo(workflows.KeywordTrigger, func(s *step) error {
+		stateStep, ok := state.Steps[s.Ref]
+		if !ok {
+			// The step not existing on the state means that it has not been processed yet.
+			// So ignore it.
+			return nil
+		}
+		statuses[s.Ref] = stateStep.Status
+		switch stateStep.Status {
+		// For each step with any of the following statuses, propagate the statuses to its dependants
+		// since they will not be executed.
+		case store.StatusErrored, store.StatusCompletedEarlyExit, store.StatusTimeout:
+			// Let's properly propagate the status to all dependents, not just direct dependents.
+			queue := []string{s.Ref}
+			for len(queue) > 0 {
+				current := queue[0] // Grab the current step reference
+				queue = queue[1:]   // Remove it from the queue
+
+				// Find the dependents for the current step reference
+				dependents, err := e.workflow.dependents(current)
+				if err != nil {
+					return err
+				}
+
+				// Propagate the status to all direct dependents
+				// With no direct dependents, it will go to the next step reference in the queue.
+				for _, sd := range dependents {
+					if _, dependentProcessed := statuses[sd.Ref]; !dependentProcessed {
+						statuses[sd.Ref] = stateStep.Status
+						// Queue the dependent for to be processed later
+						queue = append(queue, sd.Ref)
+					}
+				}
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return false, "", err
+	}
+
+	workflowProcessed := true
+	// Let's validate whether the workflow has been fully processed.
+	err = e.workflow.walkDo(workflows.KeywordTrigger, func(s *step) error {
+		// If the step is not part of the state, it is a pending step
+		// so we should consider the workflow as not fully processed.
+		if _, ok := statuses[s.Ref]; !ok {
+			workflowProcessed = false
+		}
+		return nil
+	})
+	if err != nil {
+		return false, "", err
+	}
+
+	if !workflowProcessed {
+		return workflowProcessed, "", nil
+	}
+
+	var hasErrored, hasTimedOut, hasCompletedEarlyExit bool
+	// Let's determine the status of the workflow.
+	for _, status := range statuses {
```

Comment:
I haven't read into `walkDo` suggest, but its name suggests that its walking the DAG of the workflow steps for a provided func. In this case, is the error status we present to the workflow executor the first errored status the walker finds (b/c that status is what is propagated to all the dependends of an error'd step)?

Reply:
We consider 3 states other than completed, to consider the workflow processed: error, completed_early_exit and timeout.

The way we return the error to the executor is basically as follows: if there's a single error, then it is considered error, if there's a single timeout, then it is timedout, and if none of the others, if there's a completed_early_exit, then it is considered as such.

The precedence of preference is error -> timeout -> completed_early_exit.

Not sure if this answers your question clearly, please let me know @patrickhuie19.

Reply:
To restate, these states are mutually exclusive and propagated to their dependents. If we traverse the tree in the order of executions we could return the first non-complete status we see, right?